### PR TITLE
Feature/supercedes option

### PR DESCRIPTION
**Pull request by Defkon1**
**Major changes:**
- Added `-s|--supercedes` option to `new` command
- Added `-v|--version` command

**Minor changes:**
- Added new file name sanitization method to support special characters in ADR title
- Added better exception and exit codes management
- updated CLI description
- updated README

### CODE REVIEW

**Comment by CommCody** [comment]
Hi @Defkon1,

thanks for this work, I was looking into this myself, I simply have to little time to keep working at reasonable speed.
I will take a look at this today or tomorrow.

KR,
@CommCody

**Comment by Defkon1** [comment]
Hi @CommCody !

Not a problem, take your time to review my edits and please notify me if you find something wrong/improvable.
I'm already working on new features (based on this PR) and I will send the code to you as soon they are ready...

Just one thing: I'm partecipating to Hacktoberfest 2020 and due to this year restrictions the repo must be marked by the owner to be eligible for counting towards the competition. Would you mind to add your repo the Hacktoberfest valid repo? 
(just to be clear: I will send you more PRs anyway)

**Comment by Defkon1** [comment]
Thank you very much for the label!  ;-)

File: src/adr/adr-cli.csproj [hunk 1]
```diff
@@ -4,6 +4,12 @@
     <TargetFramework>netcoreapp3.1</TargetFramework>
     <RuntimeIdentifiers>win-x64;win-x86;linux-x64;osx-x64</RuntimeIdentifiers>    
     <StartupObject>adr.Program</StartupObject>
+    <Description>A Windows equivalent of adr-tools (https://github.com/npryce/adr-tools)</Description>
+    <PackageLicenseExpression>MIT License</PackageLicenseExpression>
+    <AssemblyVersion>1.1.0.0</AssemblyVersion>
+    <FileVersion>1.1.0.0</FileVersion>
```
Comment (CommCody) [inline src/adr/adr-cli.csproj:10]: Just the `<Version>` tag is sufficient

File: src/adr/AdrEntry.cs [hunk 2]
```diff
@@ -172,10 +179,12 @@ where int.TryParse(fileNum, out fileNumOut)
         }
 
         private static string SanitizeFileName(string title)
-        {
-            return title
-                .Replace(' ', '-')
-                .ToLower();
+        {          
+            var filename = StringUtils.FoldToASCII(title.ToLower());
```
Comment (CommCody) [inline src/adr/AdrEntry.cs:183]: I'm unconvinced that this is necessary, I would imagine a lot of non English speakers would love to have their non ascii characters in a filename.

File: src/adr/Program.cs [hunk 3]
```diff
@@ -10,13 +14,20 @@ private static void Main(string[] args)
         {
             var app = new CommandLineApplication();
             app.Name = "adr-cli";
-            app.Description = "";
+            app.Description = "A Windows equivalent of adr-tools (https://github.com/npryce/adr-tools)";
```
Comment (CommCody) [inline src/adr/Program.cs]:
```suggestion
            app.Description = "A .Net equivalent of adr-tools (https://github.com/npryce/adr-tools)";
```

File: src/adr/adr-cli.csproj [hunk 4]
```diff
@@ -4,6 +4,12 @@
     <TargetFramework>netcoreapp3.1</TargetFramework>
     <RuntimeIdentifiers>win-x64;win-x86;linux-x64;osx-x64</RuntimeIdentifiers>    
     <StartupObject>adr.Program</StartupObject>
+    <Description>A Windows equivalent of adr-tools (https://github.com/npryce/adr-tools)</Description>
```
Comment (CommCody) [inline src/adr/adr-cli.csproj]:
```suggestion
    <Description>A .Net equivalent of adr-tools (https://github.com/npryce/adr-tools)</Description>
```

File: src/adr/Utils/FileUtils.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        /// <returns>a clean file name</returns>
+        internal static string SanitizeFilename(string filename, char substitution = '_')
+        {
+            filename = filename.Replace(' ', substitution);
+            string invalidCharacters = Regex.Escape(new string(System.IO.Path.GetInvalidFileNameChars()));
+            string regex = string.Format(@"([{0}]*\.+$)|([{0}]+)", invalidCharacters);
+
+            var cleanFilename = Regex.Replace(filename, regex, substitution.ToString());
+
+            var reservedWords = new[]
```
Comment (CommCody) [inline src/adr/Utils/FileUtils.cs:26]: Regarding these reserved words: I would imagine simply disallowing path serarators would be sufficient, and then always creating them inside of the given adr path?

**Review by CommCody — CHANGES_REQUESTED** [review]
I think overal good, but I want to introduce a different architecture for the application and with these changes I think we will be a lot closer

File: src/adr/AdrEntry.cs (hunk 2 above)
Comment (Defkon1) [reply src/adr/AdrEntry.cs:183]: Mmm... I don't know, I used the Apache Lucene Fold-To-ASCII approach that is quite used for cleaning filenames and to be sure about filenames consistency across different file systems...

http://opendatadpc.maps.arcgis.com/apps/opsdashboard/index.html#/b0c68bce2cce478eaac82fe38d4138b1

File: src/adr/Program.cs (hunk 3 above)
Comment (Defkon1) [reply src/adr/Program.cs]: I copied it from the repo description... ;-)

File: src/adr/adr-cli.csproj (hunk 1 above)
Comment (Defkon1) [reply src/adr/adr-cli.csproj:10]: If in future you'll want to distribute it with an MSI package (specially in old Windows versions) you'll need it when distributing an update

Comment (CommCody) [reply src/adr/adr-cli.csproj:10]: I just take my queue from [here](https://docs.microsoft.com/en-us/dotnet/core/tools/csproj#properties-per-attribute) the first note says that `Version` is the one to give all other "Version" attributes their sane default values in sdk style projects (So also `AssemblyVersion` is to go)

File: src/adr/AdrEntry.cs (hunk 2 above)
Comment (CommCody) [reply src/adr/AdrEntry.cs:183]: Ok, I've done some research and I understand you're point. I see that with the `FAT`-likes, the `ext` family and a couple of others we would run into issues.
So we will keep this

File: src/adr/Adr/AdrManager.cs [hunk 6]
```diff
@@ -0,0 +1,177 @@
+﻿using adr.Utils;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+namespace adr.Adr
+{
+    internal class AdrManager
```
Comment (CommCody) [inline src/adr/Adr/AdrManager.cs]:
```suggestion
    internal class ArchitectureDecisionLog
```

**Review by CommCody — CHANGES_REQUESTED** [review]
I had forgotten to add this comment sorry

File: src/adr/adr-cli.csproj (hunk 1 above)
Comment (Defkon1) [reply src/adr/adr-cli.csproj:10]: ok, probably this aspect is changed in .NET Core from .NET Framework

File: src/adr/Utils/FileUtils.cs (hunk 5 above)
Comment (Defkon1) [reply src/adr/Utils/FileUtils.cs:26]: yep, probably this control is excessive, also given that the structure of the file name is forced in something like `<number>_<title>.md`

the invalid file name characters should be strong enough to ensure a valid path in all os/file systems

File: src/adr/Adr/AdrManager.cs (hunk 6 above)
Comment (Defkon1) [reply src/adr/Adr/AdrManager.cs]: I'm not jealous about the name of my classes, but in this case this one is a central class that provides a lot of different methods for manage ADRs and should probably implements more and more as functionalities will be added...

I therefore believe that we should orient the class name towards something that express this function... what do you think about something like `ArchitectureDecisionProvider`?

Comment (Defkon1) [reply src/adr/Adr/AdrManager.cs]: Reading your comments on #4 I understood better your point of view on how to rename this class (giving it the "dignity" of object, and not only of an utility class), and I like it, so please forget my previous comment

**Review by Defkon1 — COMMENTED** [review]
Commented/committed all edits! ;-)

File: src/adr/Adr/AdrManager.cs (hunk 6 above)
Comment (CommCody) [reply src/adr/Adr/AdrManager.cs]: Exactly, often when you have to resort to using a "Manager"-class there is a more elegant way of expressing what its role is.

In this case you could feel that their was a need for an overarching structural element to "manage" the flow of the application. You can also see that all the operations the cli exposes operations are against the log of adr's. In the easy (already implemented) cases this involved nothing more than adding one record or creating the directory. But once you started to implement features which touched on multiple records, this naïve setup started to break down.

So it's good that you saw the need, but I think representing the cli functionalities as operations on the `ArchitectureDecisionLog` is going to allow us more flexibility in the future.

**Review by CommCody — APPROVED** [review]

**Pull request merged**